### ISSUE TITLE
For-Loop conditions evaluated multiple times

### ISSUE THREAD

**Issue by GWRon** [bmx-ng/bcc#149]
The conditions of a for loop (`for x = y to z`) are evaluated "per step"
```BlitzMax
SuperStrict
Framework Brl.Standardio

For Local i:Int = 0 To get()
	Print i
Next

Function get:Int()
	Print "get called"
	Return 3
End Function
```

NG outputs:
```
get called
0
get called
1
get called
2
get called
3
get called
```

Vanilla outputs:
```
get called
0
1
2
3
```

**Comment by GWRon** [comment]
```BlitzMax
SuperStrict
Framework Brl.Standardio

global limit:int = 3
For Local i:Int = 0 To limit
	if i = 3 then limit = 5
    Print i
Next
```

Vanilla output:
```
Executing:untitled1
0
1
2
3

Process complete
```

NG output:
```
Executing:untitled1
0
1
2
3
4
5

Process complete
```

**Comment by davecamp** [comment]
I agree with GWRon on this one.
A while ago I opened a general discussion about the 'effect' described - specifically for the for-loop. To cut a long story short - it pretty much came down to "the bounds of iterating over an 'already existing collection' should not be re-evaluated each loop unless the programmer explicitly writes code to do such."

**Comment by GWRon** [comment]
Your current solution is:

```BlitzMax

		' for functions and index access, use a new local variable
		If Not TConstExpr(TBinaryCompareExpr(expr).rhs) And Not TVarExpr(TBinaryCompareExpr(expr).rhs) And Not TMemberVarExpr(TBinaryCompareExpr(expr).rhs) Then
			Local tmp:TLocalDecl=New TLocalDecl.Create( "", TBinaryCompareExpr(expr).rhs.exprType,TBinaryCompareExpr(expr).rhs,, True )
			tmp.Semant()
			Local v:TVarExpr = New TVarExpr.Create( tmp )
			TBinaryCompareExpr(expr).rhs = New TStmtExpr.Create( New TDeclStmt.Create( tmp ), v ).Semant()
		End If
```

But I assume it should be:
```BlitzMax

		' use a new local variable for everything but constants
		If Not TConstExpr(TBinaryCompareExpr(expr).rhs) Then
			Local tmp:TLocalDecl=New TLocalDecl.Create( "", TBinaryCompareExpr(expr).rhs.exprType,TBinaryCompareExpr(expr).rhs,, True )
			tmp.Semant()
			Local v:TVarExpr = New TVarExpr.Create( tmp )
			TBinaryCompareExpr(expr).rhs = New TStmtExpr.Create( New TDeclStmt.Create( tmp ), v ).Semant()
		End If
```

Of course this is under the assumption, that "constants" are truly not modifyable with bcc-NG. If there is a way - with pointers or so, then every condition should get indirect used by the help of a local copy of the variable content.

**Comment by GWRon** [comment]
Maybe it would be a good idea to introduce a new keyword... "dynamic". 

For i = 0 to x step 1 dynamic

Means that "x" is re-evaluated on each loop cycle. This is the current behaviour...and might be intended (as it is possible with c).

I fully understand that this could be achieved with "while" or "repeat" and a custom condition.

What do you think?

**Comment by woollybah** [comment]
Says he that usually mentions "compatibility"... :pig:

**Comment by GWRon** [comment]
It would stay backwards compatible... if you do __not__ use a potential new keyword.

So I "who blames" am not fighting myself :-)


... just wanted to "use" existing flaws to out benefit. So when fixing it...you might at least comment that code block to find it easily when needed.

**Comment by GWRon** [comment]
Wonder why git does not close that issue (maybe it gets closed once you merge your changes to "master").

Ok: https://help.github.com/articles/closing-issues-via-commit-messages/
```
For example, a commit message with Fixes #45 will close issue 45 in that repository once the commit is merged into the default branch.

If the commit is in a non-default branch, the issue will remain open and the issue will be referenced with a tooltip.
```

@ pointers modifying constants
This is not (easily) possible in in vanilla, as BlitzMax blocks varptr(constValue). NG doesn't - will create an issue for this.
